### DISCUSSION TITLE
how to add to IList.RoleAssignments

### DISCUSSION THREAD

**Discussion by Ofer-Gal** [pnp/pnpcore#526] [Q&A] [answered]
Since I did not find any "IList.AddRoleAssignment" I tried the following:
`string[] rights = new string[1]{ "Contributor"};
                IList DPCDocumentsList = await pnpContext.Web.Lists.GetByTitleAsync("DPC Documents", l => l.RoleAssignments);
                IWeb web = pnpContext.Web;
                ISharePointUser sharePointUser = web.EnsureUser(grantRequest.UserEmail);
                bool done = await DPCDocumentsList.AddRoleDefinitionsAsync(sharePointUser.Id, rights);`
But I get error: 
`System.ArgumentException: Role definition 'Contributor' not found
at PnP.Core.Model.SharePoint.List.AddRoleDefinitionsAsync(Int32 principalId, String[] names)
`
I am sure I am wrong but don't know how to assign a "Contributor" right on a library with the SDK.
I will also need to have another function that will remove the "Contributor" right later.
If OK another question: How to remove taxonomy value from library field (clear existing selection with the SDK)

Thanks

**Comment by jansenbe** [comment, marked as answer]
@Ofer-Gal : hope this test (https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/ListTests.cs#L1034) provides the needed clarity. Documenting role assignments is still a pending task

**Reply by Ofer-Gal** [reply]
If I understand the test, it gives the current user admin role and add and change on Pages. This does not seem realistic. the current user should not be able to give himself permissions. :-)    
Any way I tried to give another user Contribute right like so:
`
var roleDefinition = web.RoleDefinitions.Add(roleDefName, RoleType.Contributor, new PermissionKind[] {
                    PermissionKind.AddListItems ,
                    PermissionKind.EditListItems
                });
                    ISharePointUser sharePointUser = web.EnsureUser(grantRequest.UserEmail);
                    IList DPCDocumentsList = await pnpContext.Web.Lists.GetByTitleAsync("DPC Documents", l => l.RoleAssignments);
                    if (DPCDocumentsList.AddRoleDefinitions(sharePointUser.Id, roleDefName))
`
The code worked with no error but the results was not what I expected.
I named the definition like the user (so I can easily take it away later)
So I ended with a Permission Levels with the user name but no permissions at all.

What am I doing wrong?
Is there a way to grant RoleType.Contributor directly to user id n ?

I figured it out: Just
`if (DPCDocumentsList.AddRoleDefinitions(sharePointUser.Id, "Contribute"))`
does it right

Thank you
